File: services/paper_ingestion/paper_ingestion/models/pulse.py
"""Pulse deck, card, rating, and debug Pydantic models."""

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel


class PulseCardResponse(BaseModel):
    """A single scored card within a Pulse deck."""

    card_id: int
    paper_id: int
    paper_title: str
    paper_authors: list[str]
    paper_url: str | None
    rank: int
    score: float
    llm_relevance: int | None
    llm_novelty: int | None
    reasoning: str | None
    signals: dict[str, float]
    reasoning_verified: bool | None = None
    reasoning_confidence: str | None = None
    user_state: str | None = None


class PulseDeckResponse(BaseModel):
    """A full Pulse deck for one day, including all scored cards."""

    deck_id: int
    deck_date: date
    card_count: int
    generated_at: datetime
    cards: list[PulseCardResponse]
    stats: dict
    degraded_reason: str | None = None
    is_stale: bool = False
    stale_age_days: int | None = None
    stale_diagnostics: dict | None = None
    empty_reason: str | None = None  # "no_data_yet" | None


class PulseGenerateResponse(BaseModel):
    """Response for POST /api/pulse/generate — returns job_id immediately."""

    job_id: str
    status: str


class PulseStatsResponse(BaseModel):
    """Aggregate Pulse pipeline stats over a sliding window of past runs."""

    window_days: int
    decks_generated: int
    avg_candidates: float | None
    avg_llm_calls: float | None
    avg_duration_s: float | None
    last_run_at: datetime | None
    last_error: str | None
    degraded_reason: str | None = None


class PulseRateRequest(BaseModel):
    """Body for POST /api/pulse/rate."""

    paper_id: int
    rating: Literal["up", "down", "save", "dismiss", "open"]


class PulseRateResponse(BaseModel):
    """Response for POST /api/pulse/rate."""

    status: str


class PulseExplainResponse(BaseModel):
    """Reasoning + signal breakdown for a single Pulse card."""

    card_id: int
    reasoning: str | None = None
    signals: dict = {}
    llm_relevance: float | None = None
    llm_novelty: float | None = None


class PulseDebugTopicEmbedding(BaseModel):
    key: str
    dim: int | None = None
    ok: bool
    non_null: bool


class PulseDebugTopCard(BaseModel):
    card_id: int
    paper_id: int
    title: str | None = None
    signals: dict = {}
    final_score: float
    llm_relevance: float | None = None
    llm_novelty: float | None = None


class PulseDebugResponse(BaseModel):
    """Diagnostics payload for the latest Pulse deck (GET /api/pulse/debug)."""

    deck_date: str
    card_count: int
    degraded_reason: str | None = None
    source_counts: dict = {}
    source_diagnostics: dict = {}
    topic_embeddings: list[PulseDebugTopicEmbedding] = []
    top_cards: list[PulseDebugTopCard] = []
    classifier_available: bool = False
    classifier_sample_count: int | None = None
    classifier_feature_names: list[str] = []
    classifier_auc: float | None = None
    classifier_auc_degradation_reason: str | None = None
    classifier_degradation_reason: str | None = None

File: frontend/src/types/pulse.ts
import type { LifecycleState } from './paper';

export type PulseRating = 'up' | 'down' | 'save' | 'dismiss' | 'open';

export interface PulseCardItem {
  card_id: number;
  paper_id: number;
  paper_title: string;
  paper_authors: string[];
  paper_url: string | null;
  rank: number;
  score: number;
  llm_relevance: number | null;
  llm_novelty: number | null;
  reasoning: string | null;
  reasoning_verified: boolean | null;
  reasoning_confidence: 'HIGH' | 'MEDIUM' | 'LOW' | 'UNVERIFIED' | null;
  signals: Record<string, number>;
  user_state?: LifecycleState | null;
  tags?: string[] | null;
}

export interface PulseDeck {
  deck_id: number;
  deck_date: string;
  card_count: number;
  generated_at: string;
  cards: PulseCardItem[];
  stats: Record<string, unknown>;
  degraded_reason?: string | null;
  is_stale?: boolean;
  stale_age_days?: number | null;
  stale_diagnostics?: Record<string, unknown> | null;
  empty_reason?: string | null;
}

export interface SourceHealth {
  source_type: string;
  last_request_at: string | null;
  last_success_at: string | null;
  last_status: string | null;
  cooldown_until: string | null;
  consecutive_failures: number;
}

export interface SourceRunRecord {
  source_type: string;
  started_at: string;
  finished_at: string | null;
  status: string;
  candidate_count: number;
  duration_ms: number | null;
}

export interface PulseStats {
  window_days: number;
  decks_generated: number;
  avg_candidates: number | null;
  avg_llm_calls: number | null;
  avg_duration_s: number | null;
  last_run_at: string | null;
  last_error: string | null;
  degraded_reason: string | null;
  has_learned_model?: boolean;
}

export interface WhyExplanation {
  card_id: number;
  reasoning: string | null;
  signals: Record<string, number>;
  llm_relevance: number | null;
  llm_novelty: number | null;
}

export interface PulseTopicEmbedding {
  key: string;
  dim: number | null;
  ok: boolean;
  non_null: boolean;
}

export interface PulseTopCard {
  card_id: number;
  paper_id: number;
  title: string;
  signals: Record<string, number>;
  final_score: number;
  llm_relevance: number | null;
  llm_novelty: number | null;
}

export interface PulseSourceDiagnostic {
  status: string;
  message: string;
  status_code: number | null;
  retry_after_s: number | null;
  settings_hint: string | null;
}

export interface PulseDebugInfo {
  deck_date: string;
  card_count: number;
  degraded_reason: string | null;
  source_counts: Record<string, number>;
  source_diagnostics: Record<string, PulseSourceDiagnostic>;
  topic_embeddings: PulseTopicEmbedding[];
  top_cards: PulseTopCard[];
  classifier_available: boolean;
  classifier_sample_count: number | null;
  classifier_feature_names: string[];
  classifier_auc: number | null;
  classifier_auc_degradation_reason: string | null;
  classifier_degradation_reason: string | null;
}
